### PR TITLE
Add 8,4 Lennard-Jones Pair Potential

### PR DESCRIPTION
**Pull request by jbdelony**
## Description

This PR implements an 8,4 Lennard-Jones pair potential whose functionality mirrors the standard 12,6 Lennard-Jones potential.  This is primarily implemented through the addition of `hoomd/md/EvaluatorPairLJ0804.h`. 
## Motivation and context

Despite performance improvements implemented for the Mie potential in PR #945, the Mie potential still performs approximately 2.5x worse than the equivalent Lennard-Jones potential on the GTX 1080Ti.  This PR adds an 8,4 Lennard-Jones potential which is frequently used when modeling coarse-grained polymer systems.  

## How has this been tested?

Tests have been implemented to check for invalid parameters and to validate the forces and energies of this pair potential. These tests can be found in `hoomd/md/pytest/test_potential.py`. Force and energy values for `r=0.75` and `r=1.5` have been added to `hoomd/md/pytest/forces_and_energies.json`. Benchmarks show that this pair potential performs within 2-4% of the 12,6 Lennard-Jones potential.

## Change log

```
Add 8,4 Lennard-Jones pair potential.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

[profile_LJ0804_HC_1E5.log](https://github.com/glotzerlab/hoomd-blue/files/6192451/profile_LJ0804_HC_1E5.log)
[profile_Mie_1E5.log](https://github.com/glotzerlab/hoomd-blue/files/6192452/profile_Mie_1E5.log)
[profile_LJ.log](https://github.com/glotzerlab/hoomd-blue/files/6192453/profile_LJ.log)

### CODE REVIEW

**Comment by joaander** [comment]
@jbdelony Could you try one more benchmark to confirm that the performance delta warrants maintaining this code going forward?

You are running this on a GPU that as limited double precision support. I plan to eventually implement a mixed-precision compilation #355 to boost performance on these GPUs. To test the performance of Mie in mixed precision, please modify `EvaluatorPairMie` to cast `rsq` to a float and then perform all computations in `float` precision. To be fair, make the same tweak to the 8-4 potential to compare mixed precision performance in both cases.

**Comment by jbdelony** [comment]
@joaander Sure, results are attached here.  It appears that the 8,4 LJ potential only performs ~6% better than Mie in single precision.  I also attached the modified headers (formatted as .txt) to ensure I made the appropriate cast edits.  I do have access to a Titan V, Titan RTX, and RTX 3090 if there is another GPU architecture you'd be interested in profiling.
[profile_Mie_SP_1E5.log](https://github.com/glotzerlab/hoomd-blue/files/6213586/profile_Mie_SP_1E5.log)
[profile_LJ0804_HC_SP_1E5.log](https://github.com/glotzerlab/hoomd-blue/files/6213587/profile_LJ0804_HC_SP_1E5.log)
[EvaluatorPairLJ0804.txt](https://github.com/glotzerlab/hoomd-blue/files/6213612/EvaluatorPairLJ0804.txt)
[EvaluatorPairMie.txt](https://github.com/glotzerlab/hoomd-blue/files/6213613/EvaluatorPairMie.txt)

**Comment by joaander** [comment]
Thanks! I'd be curious on the numbers for RTX 3090. That is the most recent silicon of all those GPUs.

@b-butler what do you think? You'v expressed concern in the past at the number of pair potentials we maintain. Until we get mixed precision in, this PR makes a big performance increase. After mixed precision, there is still a slight performance increase and a somewhat cleaner API for defining 8-4 LJ potentials.

**Comment by vyasr** [comment]
I have also previously expressed similar concerns about the maintenance overhead of large numbers of evaluators for different potentials. I don't think that's necessarily a reason not to do this; it seems like the performance improvement has significant impact on the science @jbdelony can produce, which should be the overriding consideration.

However, I know that recent dev meetings have included extended discussion of a minimal JIT API for HOOMD. I haven't been able to join recently (sorry about that) so I don't know what direction those discussions have gone, but this PR suggests to me that maybe we should revisit the feasibility of using JIT for MD potentials if you aren't already. Obviously there's significant work to be done to enable CUDA JIT outside HPMC patches, but since the HPMC patch code exists as an example we could build on that and hopefully take advantage of it to simplify the evaluators in the code base. Not sure if this has also been part of the recent discussions, apologies if I'm repeating.

**Comment by jbdelony** [comment]
@joaander Apologies on the delay gathering the RTX 3090 metrics, but those are attached here.  In single precision, the 8,4 LJ has a 10% performance advantage over Mie, and this advantage is roughly 250% in double precision.  In relative terms, the performance comparison is roughly equivalent between the GTX 1080Ti and RTX 3090, which I'm not sure I expected given the improved architecture.

[profile_LJ0804_1E5_RTX3090_DP.log](https://github.com/glotzerlab/hoomd-blue/files/6237686/profile_LJ0804_1E5_RTX3090_DP.log)
[profile_LJ0804_1E5_RTX3090_SP.log](https://github.com/glotzerlab/hoomd-blue/files/6237687/profile_LJ0804_1E5_RTX3090_SP.log)
[profile_Mie_1E5_RTX3090_DP.log](https://github.com/glotzerlab/hoomd-blue/files/6237688/profile_Mie_1E5_RTX3090_DP.log)
[profile_Mie_1E5_RTX3090_SP.log](https://github.com/glotzerlab/hoomd-blue/files/6237689/profile_Mie_1E5_RTX3090_SP.log)

**Comment by joaander** [comment]
> However, I know that recent dev meetings have included extended discussion of a minimal JIT API for HOOMD. I haven't been able to join recently (sorry about that) so I don't know what direction those discussions have gone, but this PR suggests to me that maybe we should revisit the feasibility of using JIT for MD potentials if you aren't already. 

We have discussed work along these lines at several meetings.

> Obviously there's significant work to be done to enable CUDA JIT outside HPMC patches, but since the HPMC patch code exists as an example we could build on that and hopefully take advantage of it to simplify the evaluators in the code base. Not sure if this has also been part of the recent discussions, apologies if I'm repeating.

The burden is not only significant on the developers, but also on users. On the CPU, the HPMC patch energy evaluators require that HOOMD be built against LLVM. On many clusters, this requires users build LLVM from source. Due to this challenge, we currently plan to offer run time compiled code capabilities as an optional component.

File: hoomd/md/EvaluatorPairLJ0804.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ0804.h:5]:
```suggestion
```
We are no longer assigning maintainers to specific files.

File: hoomd/md/EvaluatorPairLJ0804.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ0804.h:20]: We are no longer maintaining unused doxygen file level comments.
```suggestion
```

File: hoomd/md/EvaluatorPairLJ0804.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ0804.h:79]:
```suggestion
/*! EvaluatorPairLJ0804 evaluates the function:
```
There is no need to duplicate the documentation from `EvaluatorPairLJ.h`

File: hoomd/md/EvaluatorPairLJ0804.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ0804.h:89]:
```suggestion
    The LJ potential does not need diameter or charge. \a lj1 is
```
The code does not use a Scalar2.

File: hoomd/md/EvaluatorPairLJ0804.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ0804.h:130]:
```suggestion
```
There aren't any unit tests making use of this constructor.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for the careful benchmarks! We will merge this to provide the best possible performance for LJ 8-4 potentials.

I have a few suggested changes to remove duplicate or unneeded comments/code. I think these are the last changes needed before we can merge.

We have determine the next steps in enabling mixed-precision MD in #355. I plan to improve performance and validation testing first, then we can start work on mixed precision to improve performance on GPUs without many double precision units.

**Review by b-butler — APPROVED** [review]
I just looked at the Python code, but it looks good to me.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**
